python/tvm/topi/random/kernel.py

```diff
@@ -0,0 +1,407 @@
+
+    out_shape: number
+        Determines the number of ouput states to generate. :code:`state[i]` will correspond to
+        counter+i.
+    """
+    nrounds = 20
+    nwords = 4
+    iwidth = 64
+    assert nrounds % 4 == 0
+    assert nwords in [4, 8, 16]
+
+    assert key_buf.dtype == "uint64"  # TODO: support 32 bit inputs
+    assert key_buf.dtype == counter_buf.dtype
+
+    def mix(a, b, rotation):
+        x = a + b  # TODO should be wrapping
+        y = x ^ ((b << rotation) | (b >> (iwidth - rotation)))
+        return [x, y]
+
+    # temporary buffer for holding the results of _PERMUTATIONS
+    tmp = irb.allocate(out_buf.dtype, out_shape, name="tmp", scope="global")
+    tmp_offset = 0
+
+    # Initialize entire key. It is composed of the original key with one
+    # element appended. The appended element is the xor of all key words plus a
+    # constant.
+    full_key = irb.allocate("uint64", nwords + 1, name="full_key", scope="global")
+    for i in range(nwords):
+        full_key[i] = key_buf[key_offset + i]
+    # initial key constant, full_key[nwords] is equivalent to k_{N_W} in the Skein paper.
+    full_key[nwords] = tvm.tir.const(0x1BD11BDAA9FC1A22, dtype="uint64")
+    for i in range(nwords):
+        full_key[nwords] ^= key_buf[key_offset + i]  # TODO: wrapping
+
+    # TODO: overwrite counter instead?
+    with irb.for_range(0, out_shape, dtype="uint64", name="i") as i:
+        for j in range(nwords):
+            out_buf[out_offset + i * nwords + j] = counter_buf[counter_offset + j] + i
+
+    def key_schedule(s, i):
+        # Threefry uses no tweak, so the key schedule is simple
+        if i == nwords - 1:
+            return full_key[(s + i) % (nwords + 1)] + tvm.tir.const(s, dtype="uint64")
+        return full_key[(s + i) % (nwords + 1)]
+
+    with irb.for_range(0, out_shape, name="l") as l:  # pylint: disable=invalid-name
+        for i in range(nrounds // 4):
+            for j in range(nwords):
+                out_buf[out_offset + l * nwords + j] += key_schedule(i, j)  # TODO wrapping
+            for k in range(4):
+                for j in range(nwords // 2):
+                    (
+                        out_buf[out_offset + l * nwords + j * 2 + 0],
+                        out_buf[out_offset + l * nwords + j * 2 + 1],
+                    ) = mix(
+                        out_buf[out_offset + l * nwords + j * 2 + 0],
+                        out_buf[out_offset + l * nwords + j * 2 + 1],
+                        _ROTATIONS[nwords][(i * 4 + k) % 8][j],
+                    )
+                for j in range(nwords):
+                    tmp[tmp_offset + l * nwords + j] = out_buf[
+                        out_offset + l * nwords + _PERMUTATIONS[nwords][j]
+                    ]
+                # number of rounds is even, so out always contains the result
+                (out_buf, tmp) = (tmp, out_buf)
+                (out_offset, tmp_offset) = (tmp_offset, out_offset)
+
```

Comment:
I see some `TODO` in this function (`_threefry`), do they affect the correctness of the algorithm?

Reply:
The only one that matters is `TODO should be wrapping`. I do not know if TVM guarantees unsigned integer arithmetic to be wrapping (instead of saturating).